File: metaprogramming/metaprogramming-john.java
import java.util.HashMap;
import java.util.Scanner;

public class Main {
  public static void main(String[] args) {
    HashMap<String, Integer> vars = new HashMap<>();
    Scanner s = new Scanner(System.in);

    while (s.hasNextLine()) {
      String[] line = s.nextLine().split(" ");

      if (line[0].equals("define")) {
        int val = Integer.parseInt(line[1]);
        vars.put(line[2], val);
        continue;
      } else if (line[0].equals("eval")) {
        if (!vars.containsKey(line[1]) || !vars.containsKey(line[3])) {
          System.out.println("undefined");
          continue;
        }
        int a = vars.get(line[1]);
        int b = vars.get(line[3]);
        if (line[2].equals("<")) {
          System.out.println(a < b);
        } else if (line[2].equals(">")) {
          System.out.println(a > b);
        } else {
          System.out.println(a == b);
        }
      }
    }
  }
}

File: metaprogramming/metaprogramming-john.py
vars = {}

while True:
  try:
    line = input().split(" ")
    if line[0] == 'define':
      val = int(line[1])
      vars[line[2]] = val
    elif line[0] == 'eval':
      if line[1] not in vars or line[3] not in vars:
        print('undefined')
      else:
        a = vars[line[1]]
        b = vars[line[3]]
        r = None
        if line[2] == '<':
          r = a < b
        elif line[2] == '>':
          r = a > b
        else:
          r = a == b
        print('true' if r else 'false')
  except:
    break
